core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1985,6 +1992,140 @@ class KafkaApis(val requestChannel: RequestChannel,
+    def sendResponseCallback(error: Errors, expiryTimestamp: Long) {
+      trace("Sending expire token response for correlation id %d to client %s."
+        .format(request.header.correlationId, request.header.clientId))
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        new ExpireDelegationTokenResponse(requestThrottleMs, error, expiryTimestamp))
+    }
+
+    if (!allowTokenRequests(request))
+      sendResponseCallback(Errors.DELEGATION_TOKEN_REQUEST_NOT_ALLOWED, DelegationTokenManager.ErrorTimestamp)
+    else {
+      tokenManager.expireToken(
+        request.session.principal,
+        expireTokenRequest.hmac(),
+        expireTokenRequest.expiryTimePeriod(),
+        sendResponseCallback
+      )
+    }
+  }
+
+  def handleDescribeTokensRequest(request: RequestChannel.Request) {
+    val describeTokenRequest = request.body[DescribeDelegationTokenRequest]
+
+    // the callback for sending a describe token response
+    def sendResponseCallback(error: Errors, tokenDetails: List[DelegationToken]) {
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        new DescribeDelegationTokenResponse(requestThrottleMs, error, tokenDetails.asJava))
+      trace("Sending describe token response for correlation id %d to client %s."
+        .format(request.header.correlationId, request.header.clientId))
+    }
+
+    if (!allowTokenRequests(request))
+      sendResponseCallback(Errors.DELEGATION_TOKEN_REQUEST_NOT_ALLOWED, List.empty)
+    else if (!config.tokenAuthEnabled)
+      sendResponseCallback(Errors.DELEGATION_TOKEN_AUTH_DISABLED, List.empty)
+    else {
+      val requestPrincipal = request.session.principal
+
+      if (describeTokenRequest.ownersListEmpty()) {
+        sendResponseCallback(Errors.NONE, List())
+      }
+      else {
+        val owners = if (describeTokenRequest.owners == null) null else describeTokenRequest.owners.asScala
+        def eligible(token: TokenInformation) : Boolean = {
+          val include =
+          //exclude tokens which are not requested
+            if (owners != null && !owners.exists(owner => token.ownerOrRenewer(owner))) {
+              false
+              //Owners and the renewers can describe their own tokens
+            } else if (token.ownerOrRenewer(requestPrincipal)) {
+              true
+              // Check permission for non-owned tokens
+            } else if (authorize(request.session, Describe, new Resource(kafka.security.auth.DelegationToken, token.tokenId))) {
+              true
+            }
+            else {
+              false
+            }
+
+          include
+        }
+
+        val tokens =  tokenManager.getTokens(eligible)
+        sendResponseCallback(Errors.NONE, tokens)
+      }
+    }
+  }
+
+  def allowTokenRequests(request: RequestChannel.Request) : Boolean = {
+    if (request.session.principal.tokenAuthenticated || request.context.securityProtocol == SecurityProtocol.PLAINTEXT)
```

Comment:
Should we extend this to disallow requests on SSL listeners that don't enforce presenting a client certificate? Or rather, SSL endpoints with a user principal on ANONYMOUS  - otherwise even authenticated users could't request a token on such an endpoint.

Reply:
The inconsistency is that we disallow token creation on ANONYMOUS if security protocol is PLAINTEXT, but allow it if security protocol is SSL. We could probably also disable token creation if security protocol is SSL and user is ANONYMOUS.